modified the stop function in AudioPlayer

SilentCruzer opened the pull request:
## Purpose / Description
There was a bug when trying to check the pronunciation where if we use the keyboard shortcuts multiple times, the app crashes.

## Fixes
Fixes [#8637](https://github.com/ankidroid/Anki-Android/issues/8637)

## Approach
The app crashed because we were trying to record new audio while the previously recorded audio was playing. So I modified the stop() function in the AudioPlayer.java file to check the latest data source and also it resets the player after new audio is recorded.

Also added a onPrepared listener because the error - trying to use prepare() on a null object, kept crashing the app. 

## How Has This Been Tested?

Constantly press record Audio shortcut (shift + v) and play (v).
Tested this on a one plus 7t(physical device) and pixel 3A (emulator)


## Checklist
_Please, go through these checks before submitting the PR._

- [x] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code

kartikeysaran commented on AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/AudioPlayer.java:
```diff
@@ -33,17 +33,22 @@
     @Nullable private Runnable mOnStoppingListener;
     @Nullable private Runnable mOnStoppedListener;
 
-
```
Please can you revert back unnecessary changes

kartikeysaran commented on AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/AudioPlayer.java:
```diff
@@ -33,17 +33,22 @@
+                mediaPlayer.start();
+            }
+        });
+
```
Same here

kartikeysaran commented on AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/AudioPlayer.java:
```diff
@@ -63,14 +68,18 @@ private void onStopping() {
     }
 
 
+
```
Same here

kartikeysaran commented on AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/AudioPlayer.java:
```diff
@@ -63,14 +68,18 @@ private void onStopping() {
 
 
-    public void stop() {
+
```
Same here

kartikeysaran reviewed (changes requested):
Just a few small changes

kartikeysaran commented:
Can you please add a test or share a video of this ?

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/AudioPlayer.java:
```diff
@@ -33,17 +33,22 @@
         });
         mPlayer.prepare();
-        mPlayer.start();
+        mPlayer.setOnPreparedListener(new MediaPlayer.OnPreparedListener() {
```
Set the listener before calling prepare - if `.prepare()` finishes before this call, then nothing happens

mikehardy replied:
I believe this change has been made. The listener is set before prepare is called now

SilentCruzer commented:
https://user-images.githubusercontent.com/73246484/119278021-6c8c6080-bc40-11eb-9e7b-32f9c816e6c2.mp4


This is a recording of when I tested the app on the physical device.

And, I have made the mentioned changes in the latest commit.

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/AudioPlayer.java:
```diff
@@ -67,10 +72,12 @@ public void start() {
         try {
+            mPlayer.setDataSource(audioPath);
             mPlayer.prepare();
+            mPlayer.pause();
```
This doesn't seem correct - now we've moved `.start()` to `setOnPreparedListener`, we can potentially call `pause` before `.start()` is called

SilentCruzer replied:
I have added the pause() there because I needed to reset the player before a new recording starts. Looked through few forums and found that the media player in android already has a few sync issues, that's why the bug was happening. I found replacing the stop() function with the combination of pause() and seekto() functions to be effective.

david-allison-1 replied:
I don't think this is correct. 

`prepare()` calls `onPrepared` which calls `start()`, and the `pause` here seems to be working around this issue, and will fail if there's a race condition.

mikehardy reviewed (approved):
LGTM - @david-allison-1 you had comments before, if this is good for you we can merge

mikehardy reviewed (changes requested):
Oh, actually I spoke too soon, it has a compile error:

```

> Task :AnkiDroid:compilePlayDebugUnitTestKotlin
/home/runner/work/Anki-Android/Anki-Android/AnkiDroid/src/test/java/com/ichi2/anki/reviewer/KeyboardShortcutIntegrationTest.java:137: error: method stop in class AudioPlayer cannot be applied to given types;

        verify(player, times(1)).stop();
                                ^
  required: String
  found: no arguments
  reason: actual and formal argument lists differ in length
> Task :AnkiDroid:compilePlayDebugUnitTestJavaWithJavac
```

SilentCruzer commented:
> Oh, actually I spoke too soon, it has a compile error:
> 
> ```
> 
> > Task :AnkiDroid:compilePlayDebugUnitTestKotlin
> /home/runner/work/Anki-Android/Anki-Android/AnkiDroid/src/test/java/com/ichi2/anki/reviewer/KeyboardShortcutIntegrationTest.java:137: error: method stop in class AudioPlayer cannot be applied to given types;
> 
>         verify(player, times(1)).stop();
>                                 ^
>   required: String
>   found: no arguments
>   reason: actual and formal argument lists differ in length
> > Task :AnkiDroid:compilePlayDebugUnitTestJavaWithJavac
> ```

I have added the new arguments in the test since I had modified the stop function. Can you please try running the workflow again?

mikehardy commented:
Sure thing thanks @SilentCruzer - switched up the tags and pinged David to see if he had comments

mikehardy commented:
These seems silly, but a close / re-open re-runs the workflows easily so I'm doing that. Both failures in CI appeared unrelated.
